src/analysis/seminorm.lean

```diff
@@ -39,12 +43,162 @@ open_locale big_operators nnreal pointwise topological_space
+lemma sup_apply (p q : add_monoid_seminorm E) (x : E) : (p ⊔ q) x = p x ⊔ q x := rfl
+
+lemma smul_sup [has_scalar R ℝ] [has_scalar R ℝ≥0] [is_scalar_tower R ℝ≥0 ℝ]
+  (r : R) (p q : add_monoid_seminorm E) :
+  r • (p ⊔ q) = r • p ⊔ r • q :=
+have real.smul_max : ∀ x y : ℝ, r • max x y = max (r • x) (r • y),
+from λ x y, by simpa only [←smul_eq_mul, ←nnreal.smul_def, smul_one_smul ℝ≥0 r (_ : ℝ)]
+                     using mul_max_of_nonneg x y (r • 1 : ℝ≥0).prop,
+ext $ λ x, real.smul_max _ _
+
+instance : partial_order (add_monoid_seminorm E) :=
+  partial_order.lift _ fun_like.coe_injective
+
+lemma le_def (p q : add_monoid_seminorm E) : p ≤ q ↔ (p : E → ℝ) ≤ q := iff.rfl
+lemma lt_def (p q : add_monoid_seminorm E) : p < q ↔ (p : E → ℝ) < q := iff.rfl
+
+noncomputable instance : semilattice_sup (add_monoid_seminorm E) :=
+function.injective.semilattice_sup _ fun_like.coe_injective coe_sup
+
+end add_monoid_seminorm
+
 /-- A seminorm on a module over a normed ring is a function to the reals that is positive
 semidefinite, positive homogeneous, and subadditive. -/
-structure seminorm (𝕜 : Type*) (E : Type*) [semi_normed_ring 𝕜] [add_monoid E] [has_scalar 𝕜 E] :=
-(to_fun    : E → ℝ)
-(smul'     : ∀ (a : 𝕜) (x : E), to_fun (a • x) = ∥a∥ * to_fun x)
-(triangle' : ∀ x y : E, to_fun (x + y) ≤ to_fun x + to_fun y)
+structure seminorm (𝕜 : Type*) (E : Type*) [semi_normed_ring 𝕜] [add_monoid E] [has_scalar 𝕜 E]
+  extends add_monoid_seminorm E :=
+(smul' : ∀ (a : 𝕜) (x : E), to_fun (a • x) = ∥a∥ * to_fun x)
+
+lemma map_zero.of_smul {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_monoid E]
+  [smul_with_zero 𝕜 E] {f : E → ℝ} (smul : ∀ (a : 𝕜) (x : E), f (a • x) = ∥a∥ * f x) : f 0 = 0 :=
+calc f 0 = f ((0 : 𝕜) • 0) : by rw zero_smul
+     ... = 0 : by rw [smul, norm_zero, zero_mul]
+
+lemma neg.of_smul {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_comm_group E]
+  [module 𝕜 E] [norm_one_class 𝕜] {f : E → ℝ} (smul : ∀ (a : 𝕜) (x : E), f (a • x) = ∥a∥ * f x)
+  (x : E) : f (-x) = f x :=
+  calc f (-x) = f ((-1 : 𝕜) • x) : by rw neg_one_smul
+          ... = f x : by rw [smul, norm_neg, norm_one, one_mul]
+
+lemma nonneg.of {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_comm_group E] [module 𝕜 E]
+  [norm_one_class 𝕜] {f : E → ℝ} (add_le : ∀ (x y : E), f (x + y) ≤ f x + f y)
+  (smul : ∀ (a : 𝕜) (x : E), f (a • x) = ∥a∥ * f x) (x : E) : 0 ≤ f x :=
+have h: 0 ≤ 2 * f x, from
+calc 0 = f (x + (- x)) : by rw [add_neg_self, map_zero.of_smul smul]
+...    ≤ f x + f (-x)  : add_le _ _
+...    = 2 * f x : by rw [neg.of_smul smul, two_mul],
+nonneg_of_mul_nonneg_left h zero_lt_two
+
+/-- Alternative constructor for a `seminorm` on an `add_comm_group E` over a `semi_norm_ring 𝕜`
+  in which `1` has norm `1`. -/
+def seminorm.of {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_comm_group E] [module 𝕜 E]
+  [norm_one_class 𝕜] {f : E → ℝ} (add_le : ∀ (x y : E), f (x + y) ≤ f x + f y)
```

Comment:
`f` should be explicit here
```suggestion
  [norm_one_class 𝕜] (f : E → ℝ) (add_le : ∀ (x y : E), f (x + y) ≤ f x + f y)
```

Reply:
I don't mind changing it, but isn't it inferred from `add_le`?

Reply:
Compare it with `seminorm.mk`, which takes `f` implicitly.

One of your other `def`s in this PR is quite messy because you weren't able to just pass in `f`.

The other reason is so that the goal view shows `seminorm.of f _ _` instead of `seminorm.of _ _`, as the latter contains no information about the seminorm!

Reply:
Those are good points, thanks! I have made it implicit and simplified some of the defs that used it. I'll commit it when I finish with the rest of the suggestions.